src/Package/Test/DataInspect/ReplCommandsTest.cs

```diff
@@ -0,0 +1,232 @@
+            await provider.ViewObjectDetails("mtcars", null);
+
+            await viewer.Received().ViewAsync("mtcars", null);
+        }
+
+        [Test]
+        [Category.Variable.Explorer]
+        public async Task FunctionViewerTest() {
+            using (var hostScript = new RHostScript(_sessionProvider)) {
+                var funcViewer = await _aggregator.GetViewer("lm") as FunctionViewer;
+                funcViewer.Should().NotBeNull();
+
+                funcViewer.GetFileName("lm", null).Should().Be(Path.Combine(Path.GetTempPath(), "~lm.r"));
+                funcViewer.GetFileName("lm", "abc").Should().Be(Path.Combine(Path.GetTempPath(), "~abc.r"));
+
+                var code = await funcViewer.GetFunctionCode("lm");
+                code.StartsWithOrdinal("function(formula, data, subset, weights, na.action").Should().BeTrue();
+            }
+        }
+
+        [Test]
+        [Category.Variable.Explorer]
+        public async Task TableViewerTest() {
+            using (var hostScript = new RHostScript(_sessionProvider)) {
+                var gridViewer = await _aggregator.GetViewer("mtcars") as TableViewer;
+                gridViewer.Should().NotBeNull();
+
+                var eval = Substitute.For<IDebugValueEvaluationResult>();
+                eval.Classes.Returns(new List<string>() { "foo" });
+
+                gridViewer.CanView(null).Should().BeFalse();
+                gridViewer.CanView(eval).Should().BeFalse();
+
+                eval.Dim.Count.Returns(0);
+                gridViewer.CanView(eval).Should().BeFalse();
+
+                foreach (var c in new string[] { "matrix", "data.frame", "table", "array" }) {
+                    eval.Classes.Returns(new List<string>() { c });
+                    eval.Dim.Count.Returns(3);
+                    gridViewer.CanView(eval).Should().BeFalse();
+                    eval.Dim.Count.Returns(2);
+                    gridViewer.CanView(eval).Should().BeTrue();
+                    eval.Dim.Count.Returns(1);
+                    gridViewer.CanView(eval).Should().BeFalse();
+                    eval.Dim.Count.Returns(0);
+                    gridViewer.CanView(eval).Should().BeFalse();
+                }
+
+                eval.Dim.Returns((IReadOnlyList<int>)null);
+                foreach (var c in new string[] { "a", "b" }) {
+                    eval.Classes.Returns(new List<string>() { c });
+                    gridViewer.CanView(eval).Should().BeFalse();
+                }
+
+                eval.Classes.Returns(new List<string>() { "foo", "bar" });
+                gridViewer.CanView(eval).Should().BeFalse();
+            }
+        }
+
+        [Test]
+        [Category.Variable.Explorer]
+        public async Task Viewer1DTest() {
+            using (var hostScript = new RHostScript(_sessionProvider)) {
+                var viewer = await _aggregator.GetViewer("airmiles") as Viewer1D;
+                viewer.Should().NotBeNull();
```

Comment:
Yep. I will look into composite next round - there are many tests in the viewer and VE space that can be composite.